<launch>
  <arg name="image_topic"/>
  <!-- default="/camera/image"-->
  <arg name="camera_info_topic"/>
  <!--default="/camera/camera_info"-->
  <arg name="pointcloud_topic"/>
  <!--default="/pointcloud_raw"-->
  <arg name="parent_frame"/>
  <!-- The camera frame's parent, not the lidar -->
  <arg name="child_frame"/>
  <!-- The camera frame. Not the optical one-->
  <arg name="lidar_model"/>
  <!-- pandar_40p, pandar_qt-->
  <arg name="calibration_mode" default="automatic"/>

  <include file="$(find-pkg-share lidartag)/launch/lidartag_$(var lidar_model).launch.xml">
    <arg name="pointcloud_topic" value="$(var pointcloud_topic)"/>
  </include>

  <include file="$(find-pkg-share extrinsic_tag_based_calibrator)/launch/apriltag_16h5.launch.py">
    <arg name="image_topic" value="$(var image_topic)"/>
    <arg name="camera_info_topic" value="$(var camera_info_topic)"/>
  </include>

  <node pkg="extrinsic_tag_based_calibrator" exec="extrinsic_tag_based_calibrator" name="extrinsic_tag_based_calibrator_node" output="screen">
    <param name="calib_rate" value="10.0"/>
    <param name="parent_frame" value="$(var parent_frame)"/>
    <param name="child_frame" value="$(var child_frame)"/>
    <param name="base_frame" value="base_link"/>
    <param name="calibration_mode" value="$(var calibration_mode)"/>

    <!-- filtering parameters -->
    <param name="min_tag_size" value="0.6"/>
    <!--0.6 = 0.8 * 0.75 the apriltag size, not the lidartag-->
    <param name="max_tag_distance" value="20.0"/>
    <param name="max_allowed_homography_error" value="0.5"/>
    <param name="dynamics_model" value="static"/>
    <!--constant_velicity,static-->
    <param name="calibration_crossvalidation_training_ratio" value="0.7"/>
    <param name="calibration_convergence_min_pairs" value="9"/>
    <param name="calibration_convergence_min_area_percentage" value="0.1"/>
    <param name="min_pnp_points" value="8"/>
    <param name="min_convergence_time" value="6.0"/>
    <!--6.0-->
    <param name="max_no_observation_time" value="3.0"/>
    <param name="new_hypothesis_distance" value="1.5"/>
    <param name="tag_ids" value="[0, 3, 4]"/>
    <param name="tag_sizes" value="[0.6, 0.6, 0.6]"/>
    <!-- the sizes of tag_ids -->

    <param name="lidartag_max_convergence_transl" value="0.05"/>
    <param name="lidartag_max_convergence_transl_dot" value="0.03"/>
    <param name="lidartag_max_convergence_rot" value="3.0"/>
    <param name="lidartag_max_convergence_rot_dot" value="2.5"/>
    <param name="lidartag_new_hypothesis_transl" value="0.1"/>
    <param name="lidartag_new_hypothesis_rot" value="15.0"/>
    <param name="lidartag_measurement_noise_transl" value="0.05"/>
    <param name="lidartag_measurement_noise_rot" value="5.0"/>
    <param name="lidartag_process_noise_transl" value="0.01"/>
    <param name="lidartag_process_noise_transl_dot" value="0.001"/>
    <param name="lidartag_process_noise_rot" value="1.0"/>
    <param name="lidartag_process_noise_rot_dot" value="0.1"/>

    <param name="apriltag_max_convergence_transl" value="2.0"/>
    <param name="apriltag_new_hypothesis_transl" value="20.0"/>

    <param name="apriltag_measurement_noise_transl" value="0.2"/>
    <param name="apriltag_process_noise_transl" value="0.02"/>

    <remap from="image" to="$(var image_topic)"/>
    <remap from="camera_info" to="$(var camera_info_topic)"/>
  </node>
</launch>
